Origin: commit:7b6d087d463cf0b9f0f3bb3fb6a2934ef7fa6beb
Author: Evan Maddock <5157277+EbonJaeger@users.noreply.github.com>
Last-Update: 2023-01-30
Description: [PATCH 04/10] appindexer: Take into account default handlers for
 relevancy (#283)

Signed-off-by: Evan Maddock <maddock.evan@vivaldi.net>
---
 src/appindexer/Application.vala      |  2 ++
 src/appindexer/RelevancyService.vala | 21 +++++++++++++++++++++
 2 files changed, 23 insertions(+)

diff --git a/src/appindexer/Application.vala b/src/appindexer/Application.vala
index 837a3b1c..0fd29c81 100644
--- a/src/appindexer/Application.vala
+++ b/src/appindexer/Application.vala
@@ -22,6 +22,7 @@ namespace Budgie {
 		public Icon icon { get; private set; default = new ThemedIcon.with_default_fallbacks("application-default-icon"); }
 		public string desktop_path { get; private set; }
 		public string categories { get; private set; }
+		public string[] content_types { get; private set; }
 		public string generic_name { get; private set; default = ""; }
 		public bool prefers_default_gpu { get; private set; default = false; }
 		public bool should_show { get; private set; default = true; }
@@ -54,6 +55,7 @@ namespace Budgie {
 			this.desktop_path = app_info.get_filename();
 			this.keywords = app_info.get_keywords();
 			this.categories = app_info.get_categories();
+			this.content_types = app_info.get_supported_types();
 			this.generic_name = app_info.get_generic_name();
 			this.prefers_default_gpu = !app_info.get_boolean("PrefersNonDefaultGPU");
 			this.should_show = app_info.should_show();
diff --git a/src/appindexer/RelevancyService.vala b/src/appindexer/RelevancyService.vala
index fd1e325c..5dc1f16e 100644
--- a/src/appindexer/RelevancyService.vala
+++ b/src/appindexer/RelevancyService.vala
@@ -132,6 +132,13 @@ namespace Budgie {
 				}
 			}
 
+			// Check if the application is the default handler for its supported
+			// MIME types
+			if (app.name.ascii_down().has_prefix(name) && is_default_handler(app)) {
+				debug("Application '%s' is default handler", app.name);
+				score--;
+			}
+
 			// Set the score
 			scores.set(app.desktop_id, score);
 		}
@@ -193,5 +200,19 @@ namespace Budgie {
 			}
 			return false;
 		}
+
+		/**
+		 * Check if an application is the default handler for
+		 * any of its supported MIME types.
+		 */
+		private bool is_default_handler(Application app) {
+			foreach (var content_type in app.content_types) {
+				var default_app = AppInfo.get_default_for_type(content_type, false);
+				if (default_app == null) continue;
+				if (default_app.get_id() == app.desktop_id) return true;
+			}
+
+			return false;
+		}
 	}
 }
-- 
2.37.2

